engine/autocompleter.py

```diff
@@ -1,11 +1,106 @@
+"""Contains the titular Autocompleter class, which provides autocompletion suggestions."""
+from __future__ import annotations
+from functools import lru_cache
+from re import split as regex_split
 
 from .term_index import TermIndex, Term
 
+
+class _SuggestionTree:
+    """Represents Term suggestions based on a 'word list' and a TermIndex."""
+
+    term: Term
+    start_index: int
+    end_index: int
+    children: list[_SuggestionTree]
+
+    @classmethod
+    def root(cls, word_list: list[str], term_index: TermIndex) -> _SuggestionTree:
+        """Create the root of a _SuggestionTree and recursively the rest of the tree."""
+        return cls(None, 0, 0, word_list, term_index)
+
+    def __init__(
+        self,
+        term: Term,
+        start_index: int,
+        end_index: int,
+        word_list: [str],
+        term_index: TermIndex,
```

Comment:
Grouping these into a class or adding them as fields to the `_SuggestionTree` might be options to consider, I'd be interested to hear what others think.